### PR TITLE
Merging Pulse-Physiology-Engine to FHIR Integration module.

### PR DESCRIPTION
**Pull request by hopetes**
-Added Pulse-Physiology_FHIR  Module with its description.
-Modified caregiver app module, mainly in  in a way to fetch existing observations if available and modify some section of a code for optimization purpose. 
-We used effectiveDateTime to filter the latest observation in the caregiver app.

### CODE REVIEW

File: caregiver_application/__pycache__/service_config.cpython-312.pyc
Comment (PatrickBonne) [inline caregiver_application/__pycache__/service_config.cpython-312.pyc:1]: These (__pyccache__) files should not be in the archive. Please remove and adapt gitignore file

**Comment by PatrickBonne** [comment]
Why is chosen for pychache as an extra technology, while the caregiver app is in plain python ?

File: caregiver_application/__pycache__/service_config.cpython-312.pyc
Comment (PatrickBonne) [inline caregiver_application/__pycache__/service_config.cpython-312.pyc:1]: Why is chosen for pycache, as the care giver app is already in plain python. We have now one dependency more on tooling.

File: caregiver_application/cdr/cdr_fhir.py [hunk 1]
```diff
@@ -285,15 +286,24 @@ def get_patient(self, patient_id):
...
-    def get_observation(self, obsv_id):
-        try:
-            if not obsv_id.startswith("Observation/"):
-                obsv_id = "Observation/" + obsv_id
-            response = requests.get(f"{self.base_url}/{obsv_id}", headers=self.header)
-            return json.loads(response.text) if response else None
-        except:
-            return None
-
+    # def get_observation(self, obsv_id):
```
Comment (PatrickBonne) [inline caregiver_application/cdr/cdr_fhir.py:289]: The get_observation is a well defined function on the resource to get a specific resource based on ID. It is not smart top remove that from a library (cdr looks like lib)

File: caregiver_application/cdr/cdr_fhir.py [hunk 2]
```diff
@@ -285,15 +286,24 @@ def get_patient(self, patient_id):
...
+    #     try:
+    #         if not obsv_id.startswith("Observation/"):
+    #             obsv_id = "Observation/" + obsv_id
+    #         response = requests.get(f"{self.base_url}/{obsv_id}", headers=self.header)
+    #         return json.loads(response.text) if response else None
+    #     except:
+    #         return None
+    
+    # get observations for a patient and a specific code is fetching in this method, 
+    def get_observations_for_patient_and_code(self, patient_id, observation_code):
```
Comment (PatrickBonne) [inline caregiver_application/cdr/cdr_fhir.py:299]: this function is too specific for a lib: I would prefer a function like find_observations (or query_observations) that has as second parameter the fhir search string. Should be similar as 'find_patient'. However, this name is also a bit disputable.

File: caregiver_application/web_application/__init__.py [hunk 3]
```diff
@@ -3,8 +3,9 @@
 def create_app():
     app = Flask(__name__)
     app.config['SECRET_KEY'] = 'not_so_secret_key'
-
+    # from . import model
```
Comment (PatrickBonne) [inline caregiver_application/web_application/__init__.py:6]: Why to have this comment in the file ?

File: caregiver_application/web_application/templates/workflow.html [hunk 4]
```diff
@@ -121,46 +112,156 @@
...
+                }
+            }, 3000);
         }
 
-        $('#observationCode').change(()=>{
-            var dataType = $('#observationCode :selected').attr('data-type');
-            $('#hidValueType').val(dataType);
-        });       
+        // Initialize event listeners
+        function init() {
```
Comment (PatrickBonne) [inline caregiver_application/web_application/templates/workflow.html:139]: very generic name for a function that is not on the top of the file. Maybe initEvents() ?

File: caregiver_application/web_application/templates/workflow.html [hunk 5]
```diff
@@ -121,46 +112,156 @@
...
+    //   }
+    //   else if(observeDone!== undefined)
+    //     alert("Adding observation failed!");
+    //     function hideAlert(){
+    //         setTimeout(() => {
+    //         $(".alert").slideToggle()
+    //     }, 3000);
+    //     }
+
+    //     // $('#observationCode').change(()=>{
```
Comment (PatrickBonne) [inline caregiver_application/web_application/templates/workflow.html:241]: Why comments in comments ?

File: caregiver_application/web_application/views.py [hunk 6]
```diff
@@ -61,7 +59,7 @@ def patient(id):
             all_plan_defs = cdr.get_plan_definitions()
             if "name" in current_patient:
                 name = current_patient["name"][0]["given"][0] + " " + current_patient["name"][0]["family"]
-            print("all_plan_defs",all_plan_defs)
+            # print("all_plan_defs",all_plan_defs)
```
Comment (PatrickBonne) [inline caregiver_application/web_application/views.py:62]: Please remove if not needed

File: pulse-physiology_FHIR/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,66 @@
+FROM kitware/pulse:4.1.0
+
+################################################################################
+# install some python packages
+COPY requirements.txt /tmp/
+RUN pip3 install --requirement /tmp/requirements.txt
+RUN pip install bqplot
+
+# RUN jupyter nbextension enable --py bqplot
```
Comment (PatrickBonne) [inline pulse-physiology_FHIR/Dockerfile:9]: why are these not in 'requirements.txt' ?

File: workflow_capability_core/.settings/org.eclipse.core.resources.prefs [hunk 8]
```diff
@@ -0,0 +1,6 @@
+eclipse.preferences.version=1
```
Comment (PatrickBonne) [inline workflow_capability_core/.settings/org.eclipse.core.resources.prefs:1]: this should not be part of the archive.
(gitignore ?)

File: workflow_capability_core/src/main/java/com/philips/healthsuite/workflowcapability/core/wfcservice/CamundaInterface.java [hunk 9]
```diff
@@ -40,31 +41,39 @@ public String instantiateWorkflow(String workflowID, String patientID) {
...
             return null;
         }
+        System.out.println(" the response log is ==>  " +  httpResponse.getBody().getObject().get("id"));
         return (String) httpResponse.getBody().getObject().get("id");
     }
 
     @Override
     public String deployModel(File bpmnFile, String deploymentName) {
         HttpResponse<JsonNode> httpResponse = Unirest.post(camundaUrl + "/engine-rest/deployment/create")
+        // HttpResponse<JsonNode> httpResponse = Unirest.post(camundaUrl + "/engine-rest/deployment/create")
```
Comment (PatrickBonne) [inline workflow_capability_core/src/main/java/com/philips/healthsuite/workflowcapability/core/wfcservice/CamundaInterface.java:54]: ??

File: workflow_capability_core/src/main/java/com/philips/healthsuite/workflowcapability/core/knowledgemodelmanager/KnowledgeModelController.java [hunk 10]
```diff
@@ -64,7 +64,7 @@ String deployBpmnDiagram(@RequestParam("file") MultipartFile file) throws IOExce
             String fhirPlanDefinitionName = createFHIRObjects(bpmnModel);
             File modifiedFileName = addCamundaTags(bpmnModel);
             deployModel(modifiedFileName, fhirPlanDefinitionName);
-            return "Model successfully deployed";
+            return "Model successfully deployed in Camunda.";
```
Comment (PatrickBonne) [inline workflow_capability_core/src/main/java/com/philips/healthsuite/workflowcapability/core/knowledgemodelmanager/KnowledgeModelController.java:67]: Why is 'in Camunda' added on this level: here it still should be agnostic.

**Review by PatrickBonne — CHANGES_REQUESTED** [review]
Still some cleaning to do

**Review by hopetes — COMMENTED** [review]
I revert the changes in caregiver app while added the important changes on top of it without touching the old, and WFC, and system generated cache is deleted and handled in ignored-file,
There is version difference in java, currently set to java compiler 17, previously it was java compiler 11,

File: workflow_capability_core/src/main/java/com/philips/healthsuite/workflowcapability/core/wfcservice/CamundaInterface.java (hunk 9 above)
Comment (hopetes) [reply workflow_capability_core/src/main/java/com/philips/healthsuite/workflowcapability/core/wfcservice/CamundaInterface.java]: I fixed it,

File: workflow_capability_core/.settings/org.eclipse.core.resources.prefs (hunk 8 above)
Comment (hopetes) [reply workflow_capability_core/.settings/org.eclipse.core.resources.prefs:1]: I put in the ignore file

File: caregiver_application/__pycache__/service_config.cpython-312.pyc
Comment (hopetes) [reply caregiver_application/__pycache__/service_config.cpython-312.pyc:1]: I removed the files and manually and also included in git ignore file,

File: caregiver_application/cdr/cdr_fhir.py (hunk 1 above)
Comment (hopetes) [reply caregiver_application/cdr/cdr_fhir.py]: I specifically added a script to handle the effective date from database, while the existing code as it is.

File: caregiver_application/cdr/cdr_fhir.py (hunk 2 above)
Comment (hopetes) [reply caregiver_application/cdr/cdr_fhir.py:297]: The function is there to handle specific purpose, however, it is open to change if it confuses,

File: caregiver_application/web_application/__init__.py
[the remark this answers is not in the document]
Comment (hopetes) [reply caregiver_application/web_application/__init__.py:1]: I make a change to keep the original.

File: caregiver_application/web_application/__init__.py (hunk 3 above)
Comment (hopetes) [reply caregiver_application/web_application/__init__.py]: Now it is removed, I was supposed to handle the physiological model software integration before separation.

File: caregiver_application/web_application/views.py (hunk 6 above)
Comment (hopetes) [reply caregiver_application/web_application/views.py]: I removed it,

File: caregiver_application/web_application/templates/workflow.html
Comment (basbergevoet) [inline caregiver_application/web_application/templates/workflow.html:1]: If the code that is commented out in line 159 and following is no longer needed, than delete it. And then perhaps add some human readable comments to the new "change" function to explain what it does.

File: caregiver_application/web_application/views.py
Comment (basbergevoet) [inline caregiver_application/web_application/views.py:1]: Also here, keep the code clean by removing code that is commented out.

File: workflow_capability_core/.settings/org.eclipse.core.resources.prefs (hunk 8 above)
Comment (basbergevoet) [reply workflow_capability_core/.settings/org.eclipse.core.resources.prefs:1]: The file still is in the pull request however...... I guess Patrick's remark was about removing this preferences file from the commit, right?

**Review by basbergevoet — CHANGES_REQUESTED** [review]

File: workflow_capability_core/.settings/org.eclipse.core.resources.prefs (hunk 8 above)
Comment (PatrickBonne) [reply workflow_capability_core/.settings/org.eclipse.core.resources.prefs:1]: Still not removed ?

**Review by PatrickBonne — APPROVED** [review]

**Pull request merged**